File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        beforeHead = ListNode(9999, head) # [ 9999, 1, 2, 3, 4]
        right = head # [9999, right/1, 2, 3, 4]
        left = beforeHead # [left/9999, right/1, 2, 3, 4]

        while n > 0: # we'll be at [left/9999, 1, 2, right/3, 4]
            right = right.next 
            n -= 1
        
        while right:           # [left/9999, 1, 2, right/3, 4] <-- before
            right = right.next # [9999, left/1, 2, 3, right/4] <-- loop one
            left = left.next   # [9999, 1, left/2, 3, 4, right/null] <-- loop two
        
        left.next = left.next.next
        return beforeHead.next

File: Data Structures & Algorithms/reorder-linked-list/submission-1.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

public class Solution {
    public void ReorderList(ListNode head) {
        if (head==null) return;
        List<ListNode> nodes = new List<ListNode>();
        ListNode curr = head;
        while (curr!=null)
        {
            nodes.Add(curr);
            curr=curr.next;
        }
        int i=0, j=nodes.Count-1;
        while (i<j)
        {
            nodes[i++].next=nodes[j];
            if (i>=j) break;
            nodes[j--].next=nodes[i];
        }
        nodes[i].next=null;
    }
}